### PR TITLE
feat: improve UsedStateEVMInspector by adding more op-codes

### PR DESCRIPTION
**Pull request by nikoseven**
## 📝 Summary
Added support for SELFBALANCE, BALANCE, CALL, CREATE, and SELFDESTRUCT op-codes in UsedStateEVMInspector.
New test cases for evm inspector have been added.

## 💡 Motivation and Context
Related discussion in https://github.com/flashbots/rbuilder/issues/58.
For opcodes like EXTCODECOPY, EXTCODEHASH, and EXTCODESIZE, I thought we could add them in the future with a concrete use case, so that we can properly design the datastruct for it.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

### CODE REVIEW

**Comment by dvush** [comment]
@nikoseven nice, thank you. I'll look into it soon.

**Comment by ZanCorDX** [comment]
Hi!
We are considering adding some benchmarks before adding some extra weight to the execution code.
In the meantime, I see the word "Ephernal", did you mean "Ephemeral"?

**Comment by nikoseven** [comment]
> Hi! We are considering adding some benchmarks before adding some extra weight to the execution code. In the meantime, I see the word "Ephernal", did you mean "Ephemeral"?

Fixed the typo. Thanks!
I'll wait for updates after benchmark being added.

**Review by dvush — APPROVED** [review]
I don't see any issue. It looks great. One thing though, we updated reth and revm so can you please rebase. 

re additonal allocations / performance: I don't think that this changes the pictrure in any way
1. we were already instrumenting step
2. this adds call/call_end instrumentation that runs very rarely
3. we were already allocating hashmap for each slot and here it allocates a tiny bit of data that is much smaller than that (the order is just the number of internal calls)

File: crates/rbuilder/src/building/evm_inspector.rs [hunk 1]
```diff
@@ -104,9 +128,65 @@ where
...
+        outcome: revm::interpreter::CreateOutcome,
+    ) -> revm::interpreter::CreateOutcome {
+        if let Some(addr) = outcome.address {
+            self.used_state_trace.created_contracts.push(addr);
+        }
+        outcome
+    }
+
+    fn selfdestruct(&mut self, contract: Address, target: Address, value: U256) {
+        self.used_state_trace.destructed_contracts.push(contract);
```
Comment (dvush) [inline crates/rbuilder/src/building/evm_inspector.rs:176]: This should check if contract was already destroyed and return if yes. Selfdestruct can be called multiple times in one tx.

File: crates/rbuilder/src/building/evm_inspector.rs [hunk 2]
```diff
@@ -104,9 +128,65 @@ where
...
+            }
             _ => (),
         }
     }
+
+    fn call(&mut self, _: &mut EvmContext<DB>, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        if let Some(transfer_value) = inputs.transfer_value() {
+            if !transfer_value.is_zero() {
+                *self
+                    .used_state_trace
```
Comment (dvush) [inline crates/rbuilder/src/building/evm_inspector.rs:149]: This would be incorrect if call fails.

**Review by dvush — APPROVED** [review]
There is a small issue that would make balance part of the trace incorrect. I don't think that it must be fixed for this to be merged because our slot logic was not handling reverts anyway.

**Pull request merged**